### PR TITLE
ARROW-632: [Python] Add support for FixedWidthBinary type

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

File: cpp/src/arrow/python/builtin_convert.cc [hunk 1]
```diff
@@ -16,6 +16,7 @@
 // under the License.
 
 #include <Python.h>
+#include <arrow/util/logging.h>
```
Comment (wesm) [inline cpp/src/arrow/python/builtin_convert.cc]: can you include this as `"arrow/util/logging.h" with the other arrow includes? https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: cpp/src/arrow/python/builtin_convert.cc [hunk 2]
```diff
@@ -507,7 +552,13 @@ Status ConvertPySequence(PyObject* obj, MemoryPool* pool, std::shared_ptr<Array>
...
   int64_t size;
   PyDateTime_IMPORT;
-  RETURN_NOT_OK(InferArrowType(obj, &size, &type));
+  RETURN_NOT_OK(InferArrowTypeAndSize(obj, &size, &type));
+  return ConvertPySequence(obj, pool, out, type, size);
+}
+
+Status ConvertPySequence(PyObject* obj, MemoryPool* pool, std::shared_ptr<Array>* out,
+    const std::shared_ptr<DataType>& type, int64_t size) {
+  PyDateTime_IMPORT;
```
Comment (wesm) [inline cpp/src/arrow/python/builtin_convert.cc]: How about putting this in

https://github.com/apache/arrow/blob/master/cpp/src/arrow/python/config.cc#L25

`Init()` gets calls when you import pyarrow

File: cpp/src/arrow/python/builtin_convert.cc [hunk 3]
```diff
@@ -519,9 +570,27 @@ Status ConvertPySequence(PyObject* obj, MemoryPool* pool, std::shared_ptr<Array>
...
   RETURN_NOT_OK(MakeBuilder(pool, type, &builder));
   RETURN_NOT_OK(AppendPySequence(obj, type, builder));
-
   return builder->Finish(out);
 }
 
+Status ConvertPySequence(PyObject* obj, MemoryPool* pool, std::shared_ptr<Array>* out,
+    const std::shared_ptr<DataType>& type) {
+  int64_t size;
+  PyDateTime_IMPORT;
```
Comment (wesm) [inline cpp/src/arrow/python/builtin_convert.cc]: ditto

File: cpp/src/arrow/python/pandas_convert.cc [hunk 4]
```diff
@@ -176,6 +178,35 @@ Status AppendObjectStrings(StringBuilder& string_builder, PyObject** objects,
   return Status::OK();
 }
 
+Status AppendObjectFixedWidthBytes(FixedWidthBinaryBuilder& builder, PyObject** objects,
```
Comment (wesm) [inline cpp/src/arrow/python/pandas_convert.cc]: google C++ recommends putting mutable args after input args, and making them pointers, e.g. `(..., FooBuilder* sink)`

File: cpp/src/arrow/python/pandas_convert.cc [hunk 5]
```diff
@@ -236,13 +274,13 @@ class PandasConverter : public TypeVisitor {
   }
 
   Status InitNullBitmap() {
-    int null_bytes = BitUtil::BytesForBits(length_);
+    int null_bytes = static_cast<int>(BitUtil::BytesForBits(length_));
```
Comment (wesm) [inline cpp/src/arrow/python/pandas_convert.cc]: make `null_bytes` here `int64_t`

File: cpp/src/arrow/python/pandas_convert.cc [hunk 6]
```diff
@@ -358,7 +398,7 @@ inline Status PandasConverter::ConvertData(std::shared_ptr<Buffer>* data) {
 
 template <>
 inline Status PandasConverter::ConvertData<BooleanType>(std::shared_ptr<Buffer>* data) {
-  int nbytes = BitUtil::BytesForBits(length_);
+  int nbytes = static_cast<int>(BitUtil::BytesForBits(length_));
```
Comment (wesm) [inline cpp/src/arrow/python/pandas_convert.cc]: ditto

File: cpp/src/arrow/python/pandas_convert.cc [hunk 7]
```diff
@@ -419,12 +459,25 @@ Status PandasConverter::ConvertObjectStrings(std::shared_ptr<Array>* out) {
...
+  return Status::OK();
+}
+
 Status PandasConverter::ConvertBooleans(std::shared_ptr<Array>* out) {
   PyAcquireGIL lock;
 
   PyObject** objects = reinterpret_cast<PyObject**>(PyArray_DATA(arr_));
 
-  int nbytes = BitUtil::BytesForBits(length_);
+  int nbytes = static_cast<int>(BitUtil::BytesForBits(length_));
```
Comment (wesm) [inline cpp/src/arrow/python/pandas_convert.cc]: `int64_t` (BytesForBits returns this)

File: cpp/src/arrow/python/pandas_convert.cc [hunk 8]
```diff
@@ -925,6 +980,36 @@ inline Status ConvertBinaryLike(const ChunkedArray& data, PyObject** out_values)
...
 }
 
+inline Status ConvertFixedWidthBinary(const ChunkedArray& data, PyObject** out_values) {
+  PyAcquireGIL lock;
+  for (int c = 0; c < data.num_chunks(); c++) {
+    auto arr = static_cast<FixedWidthBinaryArray*>(data.chunk(c).get());
+
+    const uint8_t* data_ptr;
+    int32_t length =
+        std::dynamic_pointer_cast<FixedWidthBinaryType>(arr->type())->byte_width();
```
Comment (wesm) [inline cpp/src/arrow/python/pandas_convert.cc]: You could put this in in a little function that also does `DCHECK_EQ(Type::FIXED_WIDTH_BINARY, arr->type_enum())`

**Review by wesm — COMMENTED** [review]
Very nice, looks good modulo minor c++ nits

File: cpp/src/arrow/python/builtin_convert.cc (hunk 1 above)
Comment (cpcloud) [inline cpp/src/arrow/python/builtin_convert.cc]: Done.

File: cpp/src/arrow/python/pandas_convert.cc (hunk 4 above)
Comment (cpcloud) [inline cpp/src/arrow/python/pandas_convert.cc]: Yeah, this was following the conventions of `AppendObjectStrings`. I'll change both.

File: cpp/src/arrow/python/builtin_convert.cc (hunk 2 above)
Comment (cpcloud) [inline cpp/src/arrow/python/builtin_convert.cc]: Done.

File: cpp/src/arrow/python/builtin_convert.cc (hunk 3 above)
Comment (cpcloud) [inline cpp/src/arrow/python/builtin_convert.cc]: Done.

File: cpp/src/arrow/python/pandas_convert.cc (hunk 8 above)
Comment (cpcloud) [inline cpp/src/arrow/python/pandas_convert.cc]: I was using `DCHECK_EQ`, but it aborts which doesn't let us show the user what went wrong.

File: cpp/src/arrow/python/pandas_convert.cc (hunk 5 above)
Comment (cpcloud) [inline cpp/src/arrow/python/pandas_convert.cc]: Done

File: cpp/src/arrow/python/pandas_convert.cc (hunk 4 above)
Comment (cpcloud) [inline cpp/src/arrow/python/pandas_convert.cc]: Done

File: cpp/src/arrow/python/pandas_convert.cc (hunk 6 above)
Comment (cpcloud) [inline cpp/src/arrow/python/pandas_convert.cc]: Done

File: cpp/src/arrow/python/pandas_convert.cc (hunk 7 above)
Comment (cpcloud) [inline cpp/src/arrow/python/pandas_convert.cc]: Done.

File: cpp/src/arrow/python/pandas_convert.cc (hunk 8 above)
Comment (wesm) [inline cpp/src/arrow/python/pandas_convert.cc]: You can pipe a message into DCHECK -- DCHECK should only be used for "can't fail" errors (made by developers) though, anything else should return a proper error

Comment (cpcloud) [inline cpp/src/arrow/python/pandas_convert.cc]: Hm this is definitely a user facing error, since we're dealing with user input here.

**Comment by wesm** [comment]
Can you touch the branch to trigger a CI build (now that the Parquet stuff is hopefully fixed)?

**Comment by cpcloud** [comment]
Yep.

**Comment by cpcloud** [comment]
Green on travis, w00t.

**Comment by cpcloud** [comment]
rebasing now

**Review by wesm — APPROVED** [review]
+1. thank you for the contribution!

**Pull request closed without merging**